src/ImageAxes.jl

```diff
@@ -167,6 +178,159 @@ function Base.convert{Cdest<:Colorant,n,Csrc<:Colorant}(::Type{Array{Cdest,n}},
+function AxisArrays.axisdim{name}(S::StreamingArray, ::Type{Axis{name}})
+    isa(name, Int) && return name <= N ? name : error("axis $name greater than array dimensionality $N")
+    names = axisnames(S)
+    idx = findfirst(names, name)
+    idx == 0 && error("axis $name not found in array axes $names")
+    idx
+end
+AxisArrays.axisdim(S::StreamingArray, ax::Axis) = axisdim(S, typeof(ax))
+AxisArrays.axisdim{name,T}(S::StreamingArray, ::Type{Axis{name,T}}) = axisdim(S, Axis{name})
+
+ImageCore.nimages(S::StreamingArray) = _nimages(timeaxis(S))
+ImageCore.coords_spatial{T,N}(S::StreamingArray{T,N}) =
+    filter_space_axes(axes(S), ntuple(identity, Val{N}))
+ImageCore.spatialorder(S::StreamingArray) = filter_space_axes(axes(S), axisnames())
+ImageCore.size_spatial(img::StreamingArray)    = filter_space_axes(axes(img), size(img))
+ImageCore.indices_spatial(img::StreamingArray) = filter_space_axes(axes(img), indices(img))
+function ImageCore.assert_timedim_last(S::StreamingArray)
+    istimeaxis(axes(S)[end]) || error("time dimension is not last")
+    nothing
+end
+
+streamingaxisnames{T,N,names,P,GetIndex}(::Type{StreamingArray{T,N,names,P,GetIndex}}) =
+    names
+streamingaxisnames(S::StreamingArray) = streamingaxisnames(typeof(S))
+
+timeaxis(S::StreamingArray) = _timeaxis(axes(S)...)
+
+@inline function getindex!(dest, S::StreamingArray, axs::Axis...)
+    all(ax->isstreamedaxis(ax,S), axs) || throw(ArgumentError("$axs do not coincide with the streaming axes $(streamingaxisnames(S))"))
+    _getindex!(dest, S.getindex!, parent(S), axs...)
+end
+
+@inline function getindex!(dest, S::StreamingArray, I...)
+    axs = getslicedindices(S, I)
+    all(x->isa(x, Colon), s!filter(I, S)) || throw(ArgumentError("positional indices must use `:` for any non-streaming axes"))
+    _getindex!(dest, S.getindex!, parent(S), axs...)
+end
+
+# _getindex! just makes it easy to specialize for particular parent or function types
+@inline function _getindex!(dest, f!, P::AbstractArray, axs::Axis...)
+    v = view(P, axs...)
+    f!(dest, v)
+end
+
+function isstreamedaxis{name,T,N,saxnames}(ax::Axis{name},
+                                           S::StreamingArray{T,N,saxnames})
+    in(name, saxnames)
+end
+function is!streamedaxis{name,T,N,saxnames}(ax::Axis{name},
+                                            S::StreamingArray{T,N,saxnames})
+    !in(name, saxnames)
+end
+
+sliceindices(S::StreamingArray) = s!filter(indices(S), S)
+sliceaxes(S::StreamingArray) = s!filter(axes(S), S)
+getslicedindices(S::StreamingArray, I) = sfilter(map((ax, i) -> ax(i), axes(S), I), S)
+
+sfilter(inds, S) = _sfilter(inds, axes(S), S)
+s!filter(inds, S) = _s!filter(inds, axes(S), S)
+
+@generated function _sfilter{N}(a, axs::NTuple{N,Axis}, S::StreamingArray)
+    inds = findin(axisnames(axs.parameters...), streamingaxisnames(S))
+    Expr(:tuple, Expr[:(a[$i]) for i in inds]...)
+end
+@generated function _s!filter{N}(a, axs::NTuple{N,Axis}, S::StreamingArray)
```

Comment:
I agree, they're weird. These are the opposite axes of the ones that `sfilter` extracts, so the `!` is meant as "not." The function name can't start with `!`, otherwise it would be first.

Happy to hear better suggestions. Since it was internal I didn't sweat it too much but agree it's not pretty now.